### ISSUE TITLE
[@typescript-eslint/return-await] False positive removal of `await` leading to control flow tampering

### ISSUE THREAD

**Issue by adon-at-work** [typescript-eslint/typescript-eslint#2598]
- [v] I have tried restarting my IDE and the issue persists.
- [v] I have updated to the latest version of the packages.
- [v] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/return-await": ["error"]
  }
}
```

```TS
const typedFunctionThatReturnsPromise = async (): Promise<void> => Promise.resolve()

async function foo(ambiguouslyTypedFunctionThatReturnsPromise) {
  try {
    return await ambiguouslyTypedFunctionThatReturnsPromise()
  } catch (e) {
    console.log('if (e instanceof ImportantException) { handleImportantException() }')
  }
}

export default async (): Promise<void> => {
  try {
    console.log('functionThatMayThrow()')
    return await foo(typedFunctionThatReturnsPromise) // we did NOT expect it to flow, but all exceptions handled internally
  } catch (e) {
    console.log('disastrousOperation')
  }
}

```

```
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "declaration": true,
    "declarationMap": true,
    "esModuleInterop": true,
    "module": "commonjs",
    "moduleResolution": "node",
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "pretty": true,
    "sourceMap": true,
    "strict": true,
    "target": "es2017",
    "lib": ["es2017"]
  }
}
```

**Expected Result**

The `await` keyword inside `return await ambiguouslyTypedFunctionThatReturnsPromise()` MUST be preserved, when we applied `eslint --fix` regardless of the options, be it `always` or `in-try-catch` (default option)

**Actual Result**

The [code](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/return-await.ts#L172-L179) is unconditionally removing the keyword `await` regardless of how it is configured. In other words, the option `always` and `in-try-catch` (default option) does not help prevent the much needed `await` inside a try-catch block from removal.

With `@typescript-eslint/return-await` applied, the `eslint --fix` was found to remove the `await` keyword because of an ambiguous type that could be actually a promise (ex. `any`, which was the default type for untyped or type-unresolved functions). The code was written correctly because the developer knew `ambiguouslyTypedFunctionThatReturnsPromise()` does return a promise (it was just not perfectly typed). However, the rewritten code, without the `await` keyword, will effectively bypass any error handling logic inside the catch block, thus leaving logics such as ImportantException to be never executed. The error exception was just thrown to the outer caller's scope.

As a result, an unhandled promise exception can be unexpectedly resulted at the caller's scope, potentially leading to a server crash, bypass of important error handling, and unintended code execution.

We believe `return-await` might expect everything to be strongly-typed (despite that its README did not have such a disclaimer, nor there's any automatic check to ensure that), if not aggressively making an assumption that an untyped function (ex. `any`) could actually return a `promise`, and thus the `await` keyword must be preserved when it is inside the try-catch block for correctness. We were also surprised that even when we switched the option to `always`, the `await` keyword was still removed, as also confirmed from their source code. We also verified with latest versions of whatever packages required.

No auto fixing should tamper the control flow.

**Additional Info**

eslint didn't throw an exception. 

If we run `eslint code.ts`, the following rule error is reported
`  3:12  error  Returning an awaited value that is not a promise is not allowed                                                               @typescript-eslint/return-await`

But if we run eslint code.ts --fix`, the code is simply rewritten without any warning.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.2.0` |
| `@typescript-eslint/parser`        | `4.2.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.9.0` |
| `node`                             | `12.18.4` |

c.c. @keegan-lillo

**Comment by tadhgmister** [comment]
If it detected that the value is type `any` and make a suggestion instead of a fix would that be satisfactory? That way it still gives a warning and IDE has quick fix but the `--fix` command doesn't remove your await?

If you have a better idea of how we should handle this rule please share.

**Comment by adon-at-work** [comment]
That makes sense for the `in-try-catch` option.
Fyi. https://github.com/typescript-eslint/typescript-eslint/issues/1269 might be what caused the current behavior

Also, as @keegan-lillo once suggested to me, you might consider exposing an additional option (clearly marked as a dangerous option) to preserve the current `--fix` behavior. 

As for `always`, we expect `--fix` to always put an `await` even if it was not there initially.

Thank you @tadhgmister

**Comment by bradzacher** [comment]
Next time, please make sure you include syntactically correct and complete code as an example.
If you don't, then I have to fill in the blanks in order to test your code. In doing so there's a chance I'll fill in differently to what you intended and not be able to repro - which wastes everyones time.

---

> The code was written correctly because the developer knew ambiguouslyTypedFunctionThatReturnsPromise() does return a promise (it was just not perfectly typed)

To me this seems like the bigger problem here. Why is your codebase setup such that you can write poorly typed code like that? And why did the engineer not correctly type the code if they knew that it returned a promise?

> We believe return-await might expect everything to be strongly-typed (despite that its README did not have such a disclaimer, nor there's any automatic check to ensure that)

***ALL*** type-aware lint rules can only act on the type information that they're provided.
If you are providing poor or no types, then rules _will_ behave poorly.
You will get false-positives, false-negatives and potentially bad fixes.

There's literally nothing we can do about this - as we can only know the information you give us.
TBH I'm not sure why you'd expect anything less? You get the same behaviour from TS itself - if you have loose types then you get loose or rather no safety.

#### To put it plainly: if your codebase is loosely typed, then you can expect loose results from lint rules.

----

We could downgrade to a suggestion fixer for `any`, sure. Happy to accept a PR.

**Comment by adon-at-work** [comment]
> Next time, please make sure you include syntactically correct and complete code as an example.

I've updated the sample code to make everyone happy. I'd like to however point out that the issue was obvious by simply looking at the eslint source code that was already included.

> To me this seems like the bigger problem here. Why is your codebase setup such that you can write poorly typed code like that? And why did the engineer not correctly type the code if they knew that it returned a promise?

> To put it plainly: if your codebase is loosely typed, then you can expect loose results from lint rules.

I have to emphasis that things were actually implicitly (now, manually typed too, in my latest example) typed. The only thing not typed was the parameter at `foo`, but that's resulted no complaints from the latest typescript even with the `strict` mode being ON. And given typescript qualifies it as `strict`ly-typed without complaints, I personally hesitate to claim it's loose.

We could potentially
- blame engineers for not manually typing every single stuff, even though no warnings/errors were thrown by typescript
- blame typescript for not being advanced enough to carry along the type, despite the only caller of the function `foo` is always returning a promise
- be empathetic to accept that not everyone/everything is incredibly perfect, and hopefully understand that the bar for auto-fixing should be set high on things that are only 100% sure, rather than relatively fragile assumptions, to refrain auto-fixing from breaking others' code.

Finally, my last two cents is the rule, if desired, could be more carefully written to handle `MustBeThenable`, `MustNOTBeThenable`, and `MayBeThenable`. Nevertheless, as a security folk who're bound to writing crappy (pseudo) code, I don't consider myself as an expert in this area. So, I'm leaving patches or fixes to the more knowledgeable ones. Thank you.

**Comment by tadhgmister** [comment]
> So, I'm leaving patches or fixes to the more knowledgeable ones.

this issue has been labeled `good first issue`, and following the [issue you linked to](https://github.com/typescript-eslint/typescript-eslint/issues/1269), it's easy to find [the relevant code to check if it's `any` or `unknown`](https://github.com/typescript-eslint/typescript-eslint/pull/1270/files).  

And the logic to change something from a fix to a suggest has working examples:

https://github.com/typescript-eslint/typescript-eslint/blob/39c45f360993f9689a8460dfe4735fc4ef2331fa/packages/eslint-plugin/src/rules/no-empty-interface.ts#L86-L99

and the only part that is problematic is the unconditional fix to remove `await` if it's not `Thenable`:

https://github.com/typescript-eslint/typescript-eslint/blob/39c45f360993f9689a8460dfe4735fc4ef2331fa/packages/eslint-plugin/src/rules/return-await.ts#L172-L179

you are the first, and at the moment the only, person who wants this as a feature. So why not write the change yourself? This is open source after all!

**Comment by bradzacher** [comment]
> The only thing not typed was the parameter at foo, but that's resulted no complaints from the latest typescript even with the strict mode being ON.

I would say that you have a config issue then.
According to the config you provided, you have `strict` turned on. Turning this on also turns on `noImplicitAny`, which causes TS to error on all function parameters without explicit type annotations.

As evidenced by [the error shown in the playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBFCeAHApgEwGIFczCgS3ABUALAQygCVkpMAnSABVpAFs8JkYBeGUieHDAAUASgBcMJq3bIAPADcQeVAD5uaqWw4A6WsgggANvOSiAUGb4DgMAGbZcBMHZAghpFgCM8Ac0whMCEN4QiQ0LBx8IjJKajpGZi1kERgAbzMYOFp4NIzMmD0ael4Ad1I8WA9vPwCgkLCMByiwEnIqIoTpDnNMgF8YYHJgYmFk3PyB8ANDZG1DEB8hAHI8W1GYPEgoUhxkEDWASRZEEFptsCgAUQAPYGREZpTUmDIwVBmjk7Odq9v75tEMF6SxEeV6ZnBZmQ1y+sFQyFspEwhkq-EEogkmhkCiUqnU4yyOXSE1AkCMs3miyW9kiTlaUAAsqQQsRmCVRCC8plCvFSuVYLZXEIEChGrTom04vQIFiOCkAPTymAlTioZQwAByAHlCDBoShcBtYFAQHZ5iUADQwTyYSqGQx6v4PJwQF47d5oDYXZD0Uj2+BggZDEZCMbE-Kk6YUhbLNUQPhQZiBLUoWjkJycvoQoA), which has `noImplicitAny` turned on.

----

As mentioned by @tadhgmister (thanks for chiming in with examples!) - you're the first and only user with this issue.

As a _volunteer_ maintainer, I won't have bandwidth to work on this. So it's going to be left open for now, and unless someone else runs into this - it likely won't receive a fix.

This entire project is free and open source. If this is important to you, please consider submitting a PR.

**Comment by adon-at-work** [comment]
Thanks @tadhgmister and @bradleyayers.

@bradzacher, you're right that there's a config kind of issue. In the middle of this issue writing, I switched from running `tsc` to `tsc test.ts`, hoping to narrow down only errors from the test file. But I was truly unaware of an unknown side effect, where the `tsconfig.json` will be ignored as a consequence (and hence, falling back to the default non-strict mode). 

TIL. It was a big surprise, especially given I read `tsc --help`, as shown below. It has never mentioned the fact that "When input files are specified on the command line, the `tsconfig.json` files are ignored", which was stealthily written somewhere only at an online doc https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#using-tsconfigjson-or-jsconfigjson.

This demonstrates just yet another instance that things are perhaps sometimes built "against humanity" plus the fact that people ignorant like me can hardly explore every single line, every single flag, every single eslint rule's prerequisite that are scattered throughout the offline and online fineprints.

```
tsc --help
Version 4.0.3
Syntax:   tsc [options] [file...]

Examples: tsc hello.ts
          tsc --outFile file.js file.ts
          tsc @args.txt
          tsc --build tsconfig.json

Options:
 -h, --help                                         Print this message.
 -w, --watch                                        Watch input files.
 --pretty                                           Stylize errors and messages using color and context (experimental).
 --all                                              Show all compiler options.
 -v, --version                                      Print the compiler's version.
 --init                                             Initializes a TypeScript project and creates a tsconfig.json file.
 -p FILE OR DIRECTORY, --project FILE OR DIRECTORY  Compile the project given the path to its configuration file, or to a folder with a 'tsconfig.json'.
 -b, --build                                        Build one or more projects and their dependencies, if out of date
 -t VERSION, --target VERSION                       Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'.
 -m KIND, --module KIND                             Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'.
 --lib                                              Specify library files to be included in the compilation.
                                                      'es5' 'es6' 'es2015' 'es7' 'es2016' 'es2017' 'es2018' 'es2019' 'es2020' 'esnext' 'dom' 'dom.iterable' 'webworker' 'webworker.importscripts' 'scripthost' 'es2015.core' 'es2015.collection' 'es2015.generator' 'es2015.iterable' 'es2015.promise' 'es2015.proxy' 'es2015.reflect' 'es2015.symbol' 'es2015.symbol.wellknown' 'es2016.array.include' 'es2017.object' 'es2017.sharedmemory' 'es2017.string' 'es2017.intl' 'es2017.typedarrays' 'es2018.asyncgenerator' 'es2018.asynciterable' 'es2018.intl' 'es2018.promise' 'es2018.regexp' 'es2019.array' 'es2019.object' 'es2019.string' 'es2019.symbol' 'es2020.bigint' 'es2020.promise' 'es2020.string' 'es2020.symbol.wellknown' 'es2020.intl' 'esnext.array' 'esnext.symbol' 'esnext.asynciterable' 'esnext.intl' 'esnext.bigint' 'esnext.string' 'esnext.promise' 
 --allowJs                                          Allow javascript files to be compiled.
 --jsx KIND                                         Specify JSX code generation: 'preserve', 'react-native', or 'react'.
 -d, --declaration                                  Generates corresponding '.d.ts' file.
 --declarationMap                                   Generates a sourcemap for each corresponding '.d.ts' file.
 --sourceMap                                        Generates corresponding '.map' file.
 --outFile FILE                                     Concatenate and emit output to single file.
 --outDir DIRECTORY                                 Redirect output structure to the directory.
 --removeComments                                   Do not emit comments to output.
 --noEmit                                           Do not emit outputs.
 --strict                                           Enable all strict type-checking options.
 --noImplicitAny                                    Raise error on expressions and declarations with an implied 'any' type.
 --strictNullChecks                                 Enable strict null checks.
 --strictFunctionTypes                              Enable strict checking of function types.
 --strictBindCallApply                              Enable strict 'bind', 'call', and 'apply' methods on functions.
 --strictPropertyInitialization                     Enable strict checking of property initialization in classes.
 --noImplicitThis                                   Raise error on 'this' expressions with an implied 'any' type.
 --alwaysStrict                                     Parse in strict mode and emit "use strict" for each source file.
 --noUnusedLocals                                   Report errors on unused locals.
 --noUnusedParameters                               Report errors on unused parameters.
 --noImplicitReturns                                Report error when not all code paths in function return a value.
 --noFallthroughCasesInSwitch                       Report errors for fallthrough cases in switch statement.
 --types                                            Type declaration files to be included in compilation.
 --esModuleInterop                                  Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'.
 @<file>                                            Insert command line options and files from a file.
```

I hope I can motivate @keegan-lillo to help with putting up a PR when he's available.

**Comment by tadhgmister** [comment]
@adon-at-work  you want to always specify a `-p` or `--project` when invoking `tsc`. it's perfectly valid to do `tsc -p . FILE_TO_COMPILE.ts`

**Comment by thomasmichaelwallace** [comment]
☝️ I wanted to contribute back to this repo (as I use it all the time :)), and as the solution was pretty spelled out; this seemed a good issue to start with.

so hopefully that pr covers the 'suggest on any' solution.
